src/process/io/locks/PathLockHandler.h

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright (c) 2019 DKFZ - ODCF
+ *
+ * Distributed under the MIT License (license terms are at https://github.com/dkfz-odcf/FastqIndEx/blob/master/LICENSE.txt).
+ */
+
+#ifndef FASTQINDEX_PATHLOCKHANDLER_H
+#define FASTQINDEX_PATHLOCKHANDLER_H
+
+#include "common/ErrorAccumulator.h"
+#include "process/io/locks/LockHandler.h"
+#include <mutex>
+#include <shared_mutex>
+#include <experimental/filesystem>
+
+using namespace std;
+using std::experimental::filesystem::path;
+using std::lock_guard;
+using std::mutex;
+
+
+/**
+ * Safe (network) path lock handler class for PathSource and PathSink.
+ * Handles interprocess locking for the input/output file.
+ * Access to the index file is mutually exclusive!
+ * Reading from or writing to index files never happens with a pipe as index files are really small.
+ *
+ * @see https://stackoverflow.com/questions/12439099/interprocess-reader-writer-lock-with-boost
+ */
+class PathLockHandler : public LockHandler {
+
+private:
+
+    /**
+     * Better play safe in methods handling the above mutex.
+     * Testing this would be cumbersome, we assume that the boost developers did their job and that it works as
+     * expected.
+     */
+    mutex methodMutex;
+
+    bool readLockActive = false;
+
+    bool writeLockActive = false;
+
+    FILE* lockedFileHandle = nullptr;
+
+protected:
+
+    /**
+     * The original path of the index file.
+     */
+    path lockedFile;
+
+public:
+
+    explicit PathLockHandler(path file);
+
+    ~PathLockHandler() override;
+
+    path getIndexFile();
```

Comment:
File-locking is certainly not dependent on the content of the file. The "index" in the method name is leaking the FASTQ semantics into the locking domain.

... having said that even the idea of a FASTQ is maybe leaking into the whole program, because, as you said, the tool is useable for all regular line-based records-size text files, or not?